src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java

```diff
@@ -0,0 +1,663 @@
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.Sets;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import accord.api.RoutingKey;
+import accord.impl.DurabilityScheduling;
+import accord.impl.progresslog.DefaultProgressLog;
+import accord.impl.progresslog.TxnStateKind;
+import accord.local.CommandStores;
+import accord.local.DurableBefore;
+import accord.local.MaxConflicts;
+import accord.local.Node;
+import accord.local.RedundantBefore;
+import accord.local.RejectBefore;
+import accord.primitives.Routable;
+import accord.primitives.Status;
+import accord.primitives.Timestamp;
+import accord.primitives.Txn;
+import accord.primitives.TxnId;
+import accord.utils.Invariants;
+import accord.utils.async.AsyncChain;
+import org.apache.cassandra.cql3.statements.schema.CreateTableStatement;
+import org.apache.cassandra.db.ColumnFamilyStore;
+import org.apache.cassandra.db.DecoratedKey;
+import org.apache.cassandra.db.Keyspace;
+import org.apache.cassandra.db.marshal.ByteBufferAccessor;
+import org.apache.cassandra.db.marshal.Int32Type;
+import org.apache.cassandra.db.marshal.LongType;
+import org.apache.cassandra.db.marshal.TupleType;
+import org.apache.cassandra.db.marshal.UTF8Type;
+import org.apache.cassandra.db.marshal.UUIDType;
+import org.apache.cassandra.dht.Range;
+import org.apache.cassandra.dht.Token;
+import org.apache.cassandra.exceptions.InvalidRequestException;
+import org.apache.cassandra.schema.Schema;
+import org.apache.cassandra.schema.TableId;
+import org.apache.cassandra.schema.TableMetadata;
+import org.apache.cassandra.service.accord.AccordCommandStore;
+import org.apache.cassandra.service.accord.AccordKeyspace;
+import org.apache.cassandra.service.accord.AccordService;
+import org.apache.cassandra.service.accord.AccordStateCache;
+import org.apache.cassandra.service.accord.CommandStoreTxnBlockedGraph;
+import org.apache.cassandra.service.accord.api.AccordRoutingKey;
+import org.apache.cassandra.service.accord.api.AccordRoutingKey.TokenKey;
+import org.apache.cassandra.service.consensus.migration.ConsensusMigrationState;
+import org.apache.cassandra.service.consensus.migration.TableMigrationState;
+import org.apache.cassandra.tcm.ClusterMetadata;
+import org.apache.cassandra.utils.ByteBufferUtil;
+import org.apache.cassandra.utils.Pair;
+
+import static java.lang.String.format;
+import static java.util.Comparator.comparing;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static accord.utils.async.AsyncChains.getBlockingAndRethrow;
+import static org.apache.cassandra.schema.SchemaConstants.VIRTUAL_ACCORD_DEBUG;
+import static org.apache.cassandra.utils.ByteBufferUtil.bytes;
+
+public class AccordDebugKeyspace extends VirtualKeyspace
```

Comment:
why this view point?  the vtables that are in cep-15-accord have been for developers and operators to better understand what is going on... so "not for consumption by others" is new to me.

> and is 'unstable'
I brought this up in dev@ awhile back... we don't actually have the tools to define that right now in C* so anything that is published can't be changed... we need to address exposing tables as breakable first.

Reply:
We *cannot* provide these vtables with a stable API - they expose internal implementation details, intentionally, details we expect to change (even in a minor release). They should be members of a keyspace we can clearly document as unstable, so that users do not take a programmatic dependency upon them.

Reply:
> They should be members of a keyspace we can clearly document as unstable, so that users do not take a programmatic dependency upon them.

this is not enough to not be part of the public API.  I brought this up when finding vtables were getting added that are super brittle or with bad UX and how we can fix them without doing breaking changes.  We do not have experimental tables at the moment and we need to flesh that out as a project first if we want that.  Putting the word `debug` in the keyspace is not enough for that, and if that is a goal I am against this patch; we need to fix this in the project first.

People in dev@ are not opposed to the idea, but we need to flesh that out and document things.

Reply:
There’s a difference between poorly design UX and UX specifically documented as exposing internal details that cannot be relied upon to not change. To be clear, these tables aren’t *ever* expected to be used for programmatic access. This isn’t the same as “experimental” - this is “WE MAKE NO PROMISES EVER”.

I would also be fine with having this whole keyspace hidden behind some config property for now, so you have to enable it and read some documentation making this clear. This seems to sidestep any issue and we can revisit exposing it by default once broader discussion has happened.

Reply:
That ^.

I guess I'm fine with hiding it behind a property. Will update.

Reply:
> this is “WE MAKE NO PROMISES EVER”.

This isn't a thing in the project today, so anything exposed is now part of the public API.  This is why I am harping on this topic as its been brought up... If you want tables/keyspaces where you can break w/e you wish, we need to *add* such a thing into the project, which I strongly do not feel this patch does.

> I guess I'm fine with hiding it behind a property

If we wish to hid the new progress log tables thats fine, but I am a hard -1 on hiding the existing tables as they were designed for operators / developers to help manage a production cluster.  They are not "WE MAKE NO PROMISES EVER".

Reply:
If we’re throwing hard -1s around, I’m a hard -1 on releasing any of these as production tables, including the ones that had merged to cep15 already. They leak implementation details and are unsuitable to be treated as public APIs

Reply:
> They leak implementation details

I can't speak to migration as I don't know it well, but the 2 used tables are

1) coordination
2) txn block

for coordination it mimics the existing nodetool commands to see what CQL is running, so it shows what the txn is and how long its been coordinated.  What implementation detail are you concerned about here?

For the txn blocking one, this is a thing operators need as it shows what blocks what and why.  The API intentionally only leaks txn vs key dependencies and not much else; so what are your concerns about the implementation details?  I am also curious why you never raised these concerns when you reviewed the blocking patch; your feedback there made the table far better.

Reply:
The coordination table as discussed already I consider to be pretty useless, much less so than the one for long running CQL commands as we explicitly time these out and remove them within 10s, there is nothing that can be shown to help an operator. I did not review this table. I do absolutely support enriching the progress log table with this information (in fact, a more useful version of this, to include the phase of coordination and perhaps even the number of responses we have for that phase)

The txn block table leaked actual cache references until my recent patch, so clearly I didn't review it very carefully :/ 
But this leaks at least: SaveStatus, ExecuteAt, StoreId, but also leaks details about how we manage execution dependencies (which has already been drastically modified more than once during development).

I did not anyway honestly consider this was intended to be consumed by applications, nor was I aware of any controversy about human-only virtual tables.

Reply:
> The coordination table as discussed already I consider to be pretty useless,

stated in another comment, I am fine saying we will punt on this table closer to the release and drop it for now.  What is running is very useful but if you both don't feel we are ready for that we can drop for now.

> The txn block table... But this leaks at least: SaveStatus, ExecuteAt, StoreId, but also leaks details about how we manage execution dependencies (which has already been drastically modified more than once during development).

this table is critical for someone to answer the question "why did my txn timeout", as it shows what is blocking it still.  If we say key vs txn dependencies should not be exposed then thats mostly fine as the txn_id matters the most... if you are blocked on X you then can ask why is X blocked and keep going until you find what is stuck. But if you don't expose save status and timestamps then what can you meaningfully say why something is stuck?

So, if you prefer to drop the key vs txn part, I am fine removing in this patch. If you feel the txn id and save status is too much, then we have problems as I don't know how to answer the above question.

> I did not anyway honestly consider this was intended to be consumed by applications, nor was I aware of any controversy about human-only virtual tables.

when ever there has been a patch to alter any existing table the same conversation comes up about backwards compatibility; once it reaches the public API we have to maintain it.  We have 0 tables where this stance doesn't hold true today, this patch is arguing to be the first.

As stated in the other thread, I am fine with a view point that these tables are for testing so off by default and off in prod, but exist to aid developers writing patches or to improve test errors; there is clear value in that.